[SPMD] Support foreach optimizers with functionalization

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97897
* __->__ #97853

My first attempt was to apply the same solution as how proxy_tensor.py
handles other inplace ops. However, foreach is different in the way
that it's schema is `native_functions.yaml` does not return anything,
whereas ops like `addcmul_` and `addcdiv_` do return Tensors (Thanks
bdhirsh for teaching me this!). As a result, the proxy output
during tracing does not wrap anything, and hence we cannot correctly
connect it with subsequent operators. Modifying `native_functions.yaml`
is not a preferred solution. After discussing with bdhirsh, the
temporary solution is to do foreach functionalization as a graph
pass for now. Later, when https://github.com/pytorch/pytorch/issues/97852
is addressed, we will switch to default functionalization.

Edit: the latest version follows @bdhirsh 's suggestion on using
`make_fx` `decomposition_table` instead of implementing manual
fx.Graph tranforms to functionalize `_foreach_add_`.

mrshenli committed:
[SPMD] Support foreach optimizers with functionalization

My first attempt was to apply the same solution as how proxy_tensor.py
handles other inplace ops. However, foreach is different in the way
that it's schema is `native_functions.yaml` does not return anything,
whereas ops like `addcmul_` and `addcdiv_` do return Tensors (Thanks
@bdhirsh for teaching me this!). As a result, the proxy output
during tracing does not wrap anything, and hence we cannot correctly
connect it with subsequent operators. Modifying `native_functions.yaml`
is not a preferred solution. After discussing with @bdhirsh, the
temporary solution is to do foreach functionalization as a graph
pass for now. Later, when https://github.com/pytorch/pytorch/issues/97852
is addressed, we will switch to default functionalization.

[ghstack-poisoned]

mrshenli committed:
Update on "[SPMD] Support foreach optimizers with functionalization"


My first attempt was to apply the same solution as how proxy_tensor.py
handles other inplace ops. However, foreach is different in the way
that it's schema is `native_functions.yaml` does not return anything,
whereas ops like `addcmul_` and `addcdiv_` do return Tensors (Thanks
bdhirsh for teaching me this!). As a result, the proxy output
during tracing does not wrap anything, and hence we cannot correctly
connect it with subsequent operators. Modifying `native_functions.yaml`
is not a preferred solution. After discussing with bdhirsh, the
temporary solution is to do foreach functionalization as a graph
pass for now. Later, when https://github.com/pytorch/pytorch/issues/97852
is addressed, we will switch to default functionalization.

[ghstack-poisoned]

mrshenli commented on torch/_meta_registrations.py:
```diff
@@ -1189,6 +1189,15 @@ def meta_addbmm(self, batch1, batch2, *, beta=1, alpha=1):
     return self.new_empty(self.size())
 
 
+@register_meta([aten._foreach_add_.List])
```
Hey @bdhirsh, is this change look OK to you?

bdhirsh replied:
Looks ok to me!

One thing - the python meta impl should have the same set of its restrictions as its C++ counterpart. I also see checks that both input tensors are non-empty in the C++ version: https://github.com/pytorch/pytorch/blob/2c16b73a1b310a58a0a1e8544cb60c133e753d6f/aten/src/ATen/native/ForeachUtils.h#L40

mrshenli replied:
Good point! Thanks! Will add this.

mrshenli commented on torch/distributed/_spmd/experimental_ops.py:
```diff
@@ -17,6 +17,38 @@
 aten = torch.ops.aten  # pyre-ignore
 
 
+@register_prop_rule(aten._foreach_add.List)  # pyre-ignore
```
Hey @wanchaol is this specialization for foreach OK? If necessary, I can update `pointwise_rule` in a followup PR. It does not support list of tensors as args for now due to [this](https://github.com/pytorch/pytorch/blob/f388bec98583a31407526050900873ebe7463483/torch/distributed/_tensor/op_schema.py#L88)? 

https://github.com/pytorch/pytorch/blob/f388bec98583a31407526050900873ebe7463483/torch/distributed/_tensor/ops/common_rules.py#L248

wanchaol replied:
I think that's fine, we don't even need to update pointwise_rule as it might be non-trival to support list of inputs. I am planning on some op decomposition change that changes all of the rule registration, so things like `foreach_add` op is easier to express.

fegin commented on torch/distributed/_spmd/api.py:
```diff
@@ -204,6 +205,87 @@ def _rematerialize_optimizer(
+                )
+                getitems: Dict[fx.Node, fx.Node] = {}
+                # need another inserting_after, otherwise new ones are always
+                # inserted immediately after ``node``
```
yup, this is annoying. Should we just remove the one level of indentation of this `inserting_after` and make it have as the same as the previous `inserting_after`? The previous `inserting_after` is useless for the L265-L271.

fegin reviewed (approved):
LGTM. Left some comments.

mrshenli commented on torch/distributed/_spmd/api.py:
```diff
@@ -204,6 +205,87 @@ def _rematerialize_optimizer(
+            continue
+
+        # Only functionalize aten._foreach_* ops.
+        if node.op == OP.CALL_FUNCTION and node.target in FOREACH_OP_MAP:
```
Yep, will enable adam in the followup PR

mrshenli commented on torch/distributed/_spmd/api.py:
```diff
@@ -204,6 +205,87 @@ def _rematerialize_optimizer(
+        # If we got new arg, it means this node consumes inplace results.
+        # Replace this node with a new one that uses functionalized outputs.
+        if new_args != node.args:
+            with gm.graph.inserting_after(node):
```
good point! I thought `fx.Node` were not mutable. Let me try that

mrshenli committed:
Update on "[SPMD] Support foreach optimizers with functionalization"


My first attempt was to apply the same solution as how proxy_tensor.py
handles other inplace ops. However, foreach is different in the way
that it's schema is `native_functions.yaml` does not return anything,
whereas ops like `addcmul_` and `addcdiv_` do return Tensors (Thanks
bdhirsh for teaching me this!). As a result, the proxy output
during tracing does not wrap anything, and hence we cannot correctly
connect it with subsequent operators. Modifying `native_functions.yaml`
is not a preferred solution. After discussing with bdhirsh, the
temporary solution is to do foreach functionalization as a graph
pass for now. Later, when https://github.com/pytorch/pytorch/issues/97852
is addressed, we will switch to default functionalization.

[ghstack-poisoned]

bdhirsh commented on torch/distributed/_spmd/api.py:
```diff
@@ -314,6 +382,12 @@ def stateless_func(
+            # yet. I so implemented a mini-functionalization just for foreach
+            # ops. Remove this when the following issue is addressed.
+            # Issue: https://github.com/pytorch/pytorch/issues/97852
+            gm = _functionalize_and_reinplace_foreach_ops(gm)
```
This replaces `foreach_add_()` with `foreach_add()` + `copy_()`, right?

Are you still running into the proxy_tensor.py issue that you mentioned earlier, the first time that `foreach_add_()` is traced?

mrshenli replied:
Hey @bdhirsh 

> This replaces foreach_add_() with foreach_add() + copy_(), right?

Yep, exactly. Below is the graph after `_functionalize_and_reinplace_foreach_ops `

```
opcode         name               target                                args                                      kwargs                                          
-------------  -----------------  ------------------------------------  ----------------------------------------  ------------------------------------------------
-------------                                                                                                                                                     
placeholder    arg0_1             arg0_1                                ()                                        {}                                              
placeholder    arg0_2             arg0_2                                ()                                        {}                                              
placeholder    arg0_3             arg0_3                                ()                                        {}                                              
placeholder    arg2_1             arg2_1                                ()                                        {}                                              
placeholder    arg2_2             arg2_2                                ()                                        {}                                              
placeholder    arg3_1             arg3_1                                ()                                        {}                                              
placeholder    arg3_2             arg3_2                                ()                                        {}                                              
call_function  t                  aten.t.default                        (arg3_1,)                                 {}                                              
call_function  addmm              aten.addmm.default                    (arg3_2, arg0_3, t)                       {}                                              
call_function  sum_1              aten.sum.default                      (addmm,)                                  {}                                              
call_function  ones_like          aten.ones_like.default                (sum_1,)                                  {'pin_memory': False, 'memory_format': torch.pre
serve_format}                                                                                                                                                     
call_function  sym_size           aten.sym_size                         (arg0_3, 0)                               {}                                              
call_function  sym_size_1         aten.sym_size                         (arg3_1, 0)                               {}                                              
call_function  expand             aten.expand.default                   (ones_like, [sym_size, sym_size_1])       {}                                              
call_function  t_1                aten.t.default                        (expand,)                                 {}                                              
call_function  mm                 aten.mm.default                       (t_1, arg0_3)                             {}                                              
call_function  t_2                aten.t.default                        (mm,)                                     {}                                              
call_function  sum_2              aten.sum.dim_IntList                  (expand, [0], True)                       {}                                              
call_function  sym_size_2         aten.sym_size                         (arg3_2, 0)                               {}                                              
call_function  view               aten.view.default                     (sum_2, [sym_size_2])                     {}                                              
call_function  detach             aten.detach.default                   (view,)                                   {}
call_function  detach_1           aten.detach.default                   (detach,)                                 {}
call_function  t_3                aten.t.default                        (t_2,)                                    {}
call_function  detach_2           aten.detach.default                   (t_3,)                                    {}
call_function  detach_3           aten.detach.default                   (detach_2,)                               {}
call_function  _foreach_add_list  aten._foreach_add.List                ([arg3_1, arg3_2], [detach_3, detach_1])  {'alpha': -0.01}
call_function  getitem_1          <function getitem at 0x7f10cdeb3f70>  (_foreach_add_list, 1)                    {}
call_function  copy__1            aten.copy_                            (arg3_2, getitem_1)                       {}
call_function  getitem            <function getitem at 0x7f10cdeb3f70>  (_foreach_add_list, 0)                    {}
call_function  copy_              aten.copy_                            (arg3_1, getitem)                         {}
output         output             output                                ([None, copy_, copy__1],)                 {}
```

mrshenli replied:
> Are you still running into the proxy_tensor.py issue that you mentioned earlier, the first time that foreach_add_() is traced?

`_functionalize_and_reinplace_foreach_ops` kind of walk around this problem. Just to be clear, the original problem was that the output of `aten._foreach_add_.List` (which None) was not connected to graph output, as a result, we cannot figure out dependencies correctly. 

```
opcode         name           target                   args                                      kwargs                                                  
-------------  -------------  -----------------------  ----------------------------------------  -------------------------------------------------------------    
placeholder    arg0_1         arg0_1                   ()                                        {}                                                               
placeholder    arg0_2         arg0_2                   ()                                        {}                                                               
placeholder    arg0_3         arg0_3                   ()                                        {}                                                               
placeholder    arg2_1         arg2_1                   ()                                        {}                                                               
placeholder    arg2_2         arg2_2                   ()                                        {}                                                               
placeholder    arg3_1         arg3_1                   ()                                        {}                                                               
placeholder    arg3_2         arg3_2                   ()                                        {}                                                               
call_function  t              aten.t.default           (arg3_1,)                                 {}                                                               
call_function  addmm          aten.addmm.default       (arg3_2, arg0_3, t)                       {}                                                               
call_function  sum_1          aten.sum.default         (addmm,)                                  {}                                                               
call_function  ones_like      aten.ones_like.default   (sum_1,)                                  {'pin_memory': False, 'memory_format': torch.preserve_format}    
call_function  sym_size       aten.sym_size            (arg0_3, 0)                               {}                                                               
call_function  sym_size_1     aten.sym_size            (arg3_1, 0)                               {}                                                               
call_function  expand         aten.expand.default      (ones_like, [sym_size, sym_size_1])       {}                                                               
call_function  t_1            aten.t.default           (expand,)                                 {}                                                               
call_function  mm             aten.mm.default          (t_1, arg0_3)                             {}                                                               
call_function  t_2            aten.t.default           (mm,)                                     {}                                                               
call_function  sum_2          aten.sum.dim_IntList     (expand, [0], True)                       {}                                                               
call_function  sym_size_2     aten.sym_size            (arg3_2, 0)                               {}                                                               
call_function  view           aten.view.default        (sum_2, [sym_size_2])                     {}                                                               
call_function  detach         aten.detach.default      (view,)                                   {}                                                               
call_function  detach_1       aten.detach.default      (detach,)                                 {}                                                               
call_function  t_3            aten.t.default           (t_2,)                                    {}                                                               
call_function  detach_2       aten.detach.default      (t_3,)                                    {}                                                               
call_function  detach_3       aten.detach.default      (detach_2,)                               {}                                                               
call_function  _foreach_add_  aten._foreach_add_.List  ([arg3_1, arg3_2], [detach_3, detach_1])  {'alpha': -0.01}                                                 
output         output         output                   ([None, arg3_1, arg3_2],)                 {}   
```

bdhirsh replied:
I think there might actually be a more elegant solution here than writing this graph pass: when you run make_fx(), you can specify a decomposition table - so one option is just to include a decomposition, from `_foreach_add_() into _foreach_add() + copy_()!

```
def foreach_add_decomp(self, other, scalar=1):
    self_updated = torch.ops.aten._foreach_add.List(self, other, scalar=scalar)
    for s, s_u in zip(self, self_updated):
        s.copy_(s_u)

...

gm = make_fx(
                    partial(stateless_func, func),
                    tracing_mode="symbolic",
                    decomposition_table={torch.ops.aten._foreach_add_.List: foreach_add_decomp},
                    _allow_non_fake_inputs=False,
                )(args, kwargs, named_states, params_and_buffers)
```

mrshenli replied:
wow, this is amazing! Let me try this. Thanks!

bdhirsh replied:
> _functionalize_and_reinplace_foreach_ops kind of walk around this problem.

I only mention that because I see that you're running make_fx() first to get a graph with `_foreach_add_()`, and then you're manually running the graph pass afterwards. I thought you might be hitting the problem you mentioned in this bit of code, with proxy tensor (but I could be wrong): https://github.com/pytorch/pytorch/blob/2c16b73a1b310a58a0a1e8544cb60c133e753d6f/torch/fx/experimental/proxy_tensor.py#L359

mrshenli replied:
my bad, I didn't register meta ops correctly

mrshenli committed:
Update on "[SPMD] Support foreach optimizers with functionalization"


My first attempt was to apply the same solution as how proxy_tensor.py
handles other inplace ops. However, foreach is different in the way
that it's schema is `native_functions.yaml` does not return anything,
whereas ops like `addcmul_` and `addcdiv_` do return Tensors (Thanks
bdhirsh for teaching me this!). As a result, the proxy output
during tracing does not wrap anything, and hence we cannot correctly
connect it with subsequent operators. Modifying `native_functions.yaml`
is not a preferred solution. After discussing with bdhirsh, the
temporary solution is to do foreach functionalization as a graph
pass for now. Later, when https://github.com/pytorch/pytorch/issues/97852
is addressed, we will switch to default functionalization.

[ghstack-poisoned]

mrshenli committed:
Update on "[SPMD] Support foreach optimizers with functionalization"


My first attempt was to apply the same solution as how proxy_tensor.py
handles other inplace ops. However, foreach is different in the way
that it's schema is `native_functions.yaml` does not return anything,
whereas ops like `addcmul_` and `addcdiv_` do return Tensors (Thanks
bdhirsh for teaching me this!). As a result, the proxy output
during tracing does not wrap anything, and hence we cannot correctly
connect it with subsequent operators. Modifying `native_functions.yaml`
is not a preferred solution. After discussing with bdhirsh, the
temporary solution is to do foreach functionalization as a graph
pass for now. Later, when https://github.com/pytorch/pytorch/issues/97852
is addressed, we will switch to default functionalization.

Edit: the latest version follows bdhirsh 's suggestion on using
`make_fx` `decomposition_table` instead of implementing manual
fx.Graph tranforms to functionalize `_foreach_add_`.

[ghstack-poisoned]

mrshenli committed:
Update on "[SPMD] Support foreach optimizers with functionalization"


My first attempt was to apply the same solution as how proxy_tensor.py
handles other inplace ops. However, foreach is different in the way
that it's schema is `native_functions.yaml` does not return anything,
whereas ops like `addcmul_` and `addcdiv_` do return Tensors (Thanks
bdhirsh for teaching me this!). As a result, the proxy output
during tracing does not wrap anything, and hence we cannot correctly
connect it with subsequent operators. Modifying `native_functions.yaml`
is not a preferred solution. After discussing with bdhirsh, the
temporary solution is to do foreach functionalization as a graph
pass for now. Later, when https://github.com/pytorch/pytorch/issues/97852
is addressed, we will switch to default functionalization.

Edit: the latest version follows bdhirsh 's suggestion on using
`make_fx` `decomposition_table` instead of implementing manual
fx.Graph tranforms to functionalize `_foreach_add_`.

[ghstack-poisoned]

wanchaol reviewed (approved):
looks good. I have some basic question about why we need to care about `foreach`, I assume we don't really need the `foreach` kernel as when optimizer graph lowering to inductor it should be able to figure out the fusion to do things similar to foreach?

fegin commented:
> looks good. I have some basic question about why we need to care about `foreach`, I assume we don't really need the `foreach` kernel as when optimizer graph lowering to inductor it should be able to figure out the fusion to do things similar to foreach?

@lessw2020 has done a test with using non fused Adam and rely on Inductor to do the fusion. The performance is bad. Of course that is Adam and we didn't test against with SGD. So we don't know how good it is or how bad it is for Inductor to fuse SGD. Regardless, we have to trace `_fused_adam_` which has a similar no-output-dependency issue as `foreach`.

mrshenli committed:
Update on "[SPMD] Support foreach optimizers with functionalization"


My first attempt was to apply the same solution as how proxy_tensor.py
handles other inplace ops. However, foreach is different in the way
that it's schema is `native_functions.yaml` does not return anything,
whereas ops like `addcmul_` and `addcdiv_` do return Tensors (Thanks
bdhirsh for teaching me this!). As a result, the proxy output
during tracing does not wrap anything, and hence we cannot correctly
connect it with subsequent operators. Modifying `native_functions.yaml`
is not a preferred solution. After discussing with bdhirsh, the
temporary solution is to do foreach functionalization as a graph
pass for now. Later, when https://github.com/pytorch/pytorch/issues/97852
is addressed, we will switch to default functionalization.

Edit: the latest version follows bdhirsh 's suggestion on using
`make_fx` `decomposition_table` instead of implementing manual
fx.Graph tranforms to functionalize `_foreach_add_`.

[ghstack-poisoned]

mrshenli committed:
Update on "[SPMD] Support foreach optimizers with functionalization"


My first attempt was to apply the same solution as how proxy_tensor.py
handles other inplace ops. However, foreach is different in the way
that it's schema is `native_functions.yaml` does not return anything,
whereas ops like `addcmul_` and `addcdiv_` do return Tensors (Thanks
bdhirsh for teaching me this!). As a result, the proxy output
during tracing does not wrap anything, and hence we cannot correctly
connect it with subsequent operators. Modifying `native_functions.yaml`
is not a preferred solution. After discussing with bdhirsh, the
temporary solution is to do foreach functionalization as a graph
pass for now. Later, when https://github.com/pytorch/pytorch/issues/97852
is addressed, we will switch to default functionalization.

Edit: the latest version follows bdhirsh 's suggestion on using
`make_fx` `decomposition_table` instead of implementing manual
fx.Graph tranforms to functionalize `_foreach_add_`.

[ghstack-poisoned]

mrshenli commented:
Test failure is irrelevant

```
cat: /Users/runner/Library/Logs/scan/*.log: No such file or directory
```

mrshenli committed:
Update on "[SPMD] Support foreach optimizers with functionalization"


My first attempt was to apply the same solution as how proxy_tensor.py
handles other inplace ops. However, foreach is different in the way
that it's schema is `native_functions.yaml` does not return anything,
whereas ops like `addcmul_` and `addcdiv_` do return Tensors (Thanks
bdhirsh for teaching me this!). As a result, the proxy output
during tracing does not wrap anything, and hence we cannot correctly
connect it with subsequent operators. Modifying `native_functions.yaml`
is not a preferred solution. After discussing with bdhirsh, the
temporary solution is to do foreach functionalization as a graph
pass for now. Later, when https://github.com/pytorch/pytorch/issues/97852
is addressed, we will switch to default functionalization.

Edit: the latest version follows bdhirsh 's suggestion on using
`make_fx` `decomposition_table` instead of implementing manual
fx.Graph tranforms to functionalize `_foreach_add_`.

[ghstack-poisoned]
